### PR TITLE
Restrict annotation value to true, map<anydata|readonly> , map<anydata|readonly>[]

### LINKED ISSUE #15533: [Annotations] Restrict constraints of maps/fields of records to be anydata and freeze the values
**Description:**
According to the v2019R1 spec, 

```
annotation-decl :=
   metadata
   [public] [const] annotation [type-descriptor] annot-tag 
   [on annot-attach-points]
annot-tag := identifier

The type-descriptor specifies the type of the annotation tag. The type must be 
a subtype of one of the following three types: true, map<anydata>, map<anydata>[].

If the annotation-decl for a tag specifies const, then a mapping-constructor-expr 
in annotations with that tag must be a const-expr and is evaluated at compile-time 
with the semantics of a const-expr. Otherwise, the mapping-constructor-expr is 
evaluated when the annotation is evaluated and the resulting value is automatically 
frozen.
```

In the initial iteration to introduce annotation changes the following is not done:
- restricting the constraints of maps/fields of records to be `anydata` - cannot be done at the moment since annotations currently have fields of types `service` (websocket), `function` (transactions), etc.
- freezing the annot value - cannot be done since the value may have non-anydata values

**Related Issues:**
https://github.com/ballerina-platform/ballerina-spec/issues/70

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
Fixes #15533

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: stdlib/grpc/src/main/ballerina/src/grpc/annotation.bal [hunk 1]
```diff
@@ -59,7 +59,7 @@ public annotation GrpcResourceConfig ResourceConfig on resource function;
 # + descMap - Service dependent descriptor map, which should be set at the compile time
 public type ServiceDescriptorData record {|
     string descriptor = "";
-    map<any> descMap = {};
+    map<anydata> descMap = {};
```
Comment (KRVPerera) [inline stdlib/grpc/src/main/ballerina/src/grpc/annotation.bal:62]: @daneshk please help review if this change is okay.

Comment (daneshk) [reply stdlib/grpc/src/main/ballerina/src/grpc/annotation.bal:62]: yes, it's okay

File: stdlib/messaging/rabbitmq/src/main/ballerina/src/rabbitmq/rabbitmq_commons.bal [hunk 2]
```diff
@@ -68,7 +68,7 @@ public type QueueConfiguration record {|
     boolean durable = false;
     boolean exclusive = false;
     boolean autoDelete = true;
-    map<any>? arguments = ();
+    map<anydata>? arguments = ();
```
Comment (KRVPerera) [inline stdlib/messaging/rabbitmq/src/main/ballerina/src/rabbitmq/rabbitmq_commons.bal:71]: @aashikam Would you please check changes on `rabbitmq_commons.bal`

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/annotations/AnnotationAttachmentsCompletionTest.java [hunk 3]
```diff
@@ -47,12 +47,12 @@ public void loadAnnotationCache() {
 //                {"annotationAccessExpression4.json", "annotation"},
                 {"annotationAccessExpression5.json", "annotation"},
                 {"annotationAccessExpression6.json", "annotation"},
-                {"annotationBodyCompletion1.json", "annotation"},
-                {"annotationBodyCompletion2.json", "annotation"},
+//                {"annotationBodyCompletion1.json", "annotation"},
```
Comment (KRVPerera) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/annotations/AnnotationAttachmentsCompletionTest.java:50]: Tracking issue: #24131

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 4]
```diff
@@ -608,11 +608,12 @@ public void visit(BLangAnnotationAttachment astAnnotAttach) {
     }
 
     private boolean isCompileTimeAnnotationValue(BLangExpression expr) {
-        // TODO Compile time literal constants
         switch (expr.getKind()) {
             case LITERAL:
             case NUMERIC_LITERAL:
                 return true;
+            case INVOCATION:
+                return isCompileTimeAnnotationValue(((BLangInvocation) expr).expr);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:616]: Shouldn't we pass `(BLangInvocation) expr` itself to `isCompileTimeAnnotationValue` and add a case there to check the return type?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 5]
```diff
@@ -668,6 +669,8 @@ private BIRAnnotationValue createAnnotationValue(BLangExpression expr) {
                 return createAnnotationValue(((BLangTypeConversionExpr) expr).expr);
             case STATEMENT_EXPRESSION:
                 return createAnnotationRecordValue((BLangStatementExpression) expr);
+            case INVOCATION:
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:672]: Isn't this method only called for constants? What kind of invocations do we get here for? I assume the return type has to be a constant?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 6]
```diff
@@ -1329,17 +1329,15 @@ private boolean isValidAnnotationType(BType type) {
...
+            case TypeTags.RECORD:
+                return types.isAssignable(type, symTable.anydataOrReadOnlyMapType);
             case TypeTags.ARRAY:
                 BType elementType = ((BArrayType) type).eType;
-                return (elementType.tag == TypeTags.MAP || elementType.tag == TypeTags.RECORD) &&
-                        isValidAnnotationType(elementType);
+                if (elementType.tag == TypeTags.MAP || elementType.tag == TypeTags.RECORD) {
+                    return types.isAssignable(elementType, symTable.anydataOrReadOnlyMapType);
+                }
+                return false;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1340]: Instead, should we do 
```suggestion
                return isValidAnnotationType(elementType);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java (hunk 5 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:672]: Here also, I'm not sure if working with `((BLangInvocation) expr).expr` is the correct approach.

File: stdlib/auth/src/main/ballerina/src/auth/inbound_basic_auth_provider.bal [hunk 7]
```diff
@@ -36,12 +36,12 @@ public type InboundBasicAuthProvider object {
 
     *InboundAuthProvider;
 
-    public BasicAuthConfig basicAuthConfig;
+    public readonly BasicAuthConfig basicAuthConfig;
```
Comment (MaryamZi) [inline stdlib/auth/src/main/ballerina/src/auth/inbound_basic_auth_provider.bal:39]: Thinking out loud, as to what approach we should follow by default.

I think we should make `InboundBasicAuthProvider` a `readonly object` instead. Reasons being the type definition will mandate only valid possibly immutable types are allowed for fields, and developers won't introduce non-readonly fields by mistake.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java [hunk 8]
```diff
@@ -174,4 +179,25 @@ public void testAnnotAvailabilty() {
...
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = "^error: \\{.*\\}InvalidUpdate message=Invalid update " +
+                  "of record field: modification not allowed on readonly value.*")
+    public void testAnnotationOnObjectTypeReadonlyValueEdit() {
+        BRunUtil.invoke(resultAccessNegative, "testAnnotationOnObjectTypeReadonlyValueEdit");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = "^error: \\{.*\\}InvalidUpdate message=Invalid update " +
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java:198]: We can assert these also in Ballerina itself, we can maybe change these later.

See https://github.com/ballerina-platform/ballerina-lang/blob/02b2ec6c18ce25569748d7f904055a2ab03f85a4/tests/jballerina-unit-test/src/test/resources/test-src/expressions/mappingconstructor/spread_op_field.bal#L149

File: stdlib/http/src/main/ballerina/src/http/auth/basic_auth_handler.bal [hunk 9]
```diff
@@ -29,7 +29,7 @@ public type BasicAuthHandler object {
     # Initializes the `BasicAuthHandler` object.
     #
     # + authProvider - The `auth:InboundAuthProvider` instance or the `auth:OutboundAuthProvider` instance
-    public function init(auth:InboundAuthProvider|auth:OutboundAuthProvider authProvider) {
+    public function init(readonly & (auth:InboundAuthProvider|auth:OutboundAuthProvider) authProvider) {
```
Comment (MaryamZi) [inline stdlib/http/src/main/ballerina/src/http/auth/basic_auth_handler.bal:32]: Shall we stick to either `T & readonly` or `readonly & T` uniformly?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java (hunk 5 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:672]: This happens due to desugar changes invocation is `cloneReadOnly`

File: stdlib/auth/src/main/ballerina/src/auth/inbound_basic_auth_provider.bal (hunk 7 above)
Comment (KRVPerera) [reply stdlib/auth/src/main/ballerina/src/auth/inbound_basic_auth_provider.bal:39]: Yes, `readonly object` is better and cleaner. Will push changes.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java (hunk 8 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java:198]: shall we do this later after error ctor changes as well?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java (hunk 5 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:672]: I am trying to go through to actual expression we wrap by `cloneReadOnly`. I think it is safe to say that `INVOCATION` is not supported.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:690]: I think here and in `isCompileTimeAnnotationValue` we should specifically check if the invocation name is `cloneReadOnly` and do what you've done only if that is the case.

File: stdlib/http/src/main/ballerina/src/http/annotation.bal [hunk 10]
```diff
@@ -137,9 +137,9 @@ public type WebSocketUpgradeConfig record {|
 # be successfully authorized. An array consisting of arrays is used to indicate that at least one scope from the sub-arrays 
 # should be successfully authorized.
 public type ServiceAuth record {|
-    boolean enabled = true;
-    InboundAuthHandlers authHandlers?;
-    Scopes scopes?;
+    readonly boolean enabled = true;
```
Comment (MaryamZi) [inline stdlib/http/src/main/ballerina/src/http/annotation.bal:140]: We don't need to add `readonly` for `anydata` fields right?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/IconAnnotationTest.java [hunk 11]
```diff
@@ -51,19 +55,22 @@ public void setup() {
...
+                "/fooIconPath.icon}");
     }
 
     @Test
     public void testIconOnObjectAndMemberFunction() {
         TypeDefinition objType = result.getAST().getTypeDefinitions().get(0);
         List<? extends AnnotationAttachmentNode> objAnnot = objType.getAnnotationAttachments();
         Assert.assertEquals(objAnnot.size(), 1);
-        Assert.assertEquals(objAnnot.get(0).getExpression().toString(), " {path: /barIconPath.icon}");
+        Assert.assertEquals(objAnnot.get(0).getExpression().toString(), " {path: /barIconPath.icon}.cloneReadOnly()");
```
Comment (KRVPerera) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/IconAnnotationTest.java:67]: Updated new test cases. Need a helper function later to get the exact expression value.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java (hunk 5 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:690]: Shall we track this by an issue.. ?

Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:690]: Fixing here
https://github.com/ballerina-platform/ballerina-lang/pull/24179

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 12]
```diff
@@ -1360,22 +1363,81 @@ private boolean isValidAnnotationType(BType type) {
...
+
+                return isAnyDataOrReadOnlyTypeSkippingObjectType(recordRestType);
+            case TypeTags.ARRAY:
+                return isValidAnnotationType(((BArrayType) type).eType);
+        }
+
+        return types.isAssignable(type, symTable.trueType);
+    }
+
+    private boolean isAnyDataOrReadOnlyTypeSkippingObjectType(BType type) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1390]: Shall we create an issue to track fixing std lib deviations and add back the object validation to annotations?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 13]
```diff
@@ -1360,22 +1363,81 @@ private boolean isValidAnnotationType(BType type) {
...
+                }
+
+                BType recordRestType = recordType.restFieldType;
+                if (recordRestType == null || recordRestType == symTable.noType) {
+                    return true;
+                }
+
+                return isAnyDataOrReadOnlyTypeSkippingObjectType(recordRestType);
+            case TypeTags.ARRAY:
+                return isValidAnnotationType(((BArrayType) type).eType);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1384]: We need to keep the part of the original check where we first check if the member type is a map or a record (https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java#L1372) or we will be allowing an array of `true` also.

We should ideally have a negative test for the same.

**Comment by shafreenAnfar** [comment]
Closed and re-opened because the github workflow has been updated.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 13 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Fixed : 0d31fb64b78458d94895d4c1fdbb91c53688f09d

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 12 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1394]: Issue #24217

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/TypedescValueImpl.java [hunk 14]
```diff
@@ -107,6 +107,11 @@ public Object copy(Map<Object, Object> refs) {
         return this;
     }
 
+    @Override
+    public void freezeDirect() {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/TypedescValueImpl.java:111]: Please create an issue to track verifying these changes with @irshadnilam's changes.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java (hunk 8 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java:198]: Please create an issue for this too?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java [hunk 15]
```diff
@@ -66,6 +68,9 @@ public void setup() {
 
         resultFour = BCompileUtil.compile("test-src/annotations/annot_availability.bal");
         Assert.assertEquals(resultFour.getErrorCount(), 0);
+
+        resultAccessNegative = BCompileUtil.compile("test-src/annotations/annotation_access_negative" +
+                                                                          ".bal");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/annotations/AnnotationRuntimeTest.java:73]: Can go on one line.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 12 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1394]: Created an issue to do language changes after StandardLib changes: https://github.com/ballerina-platform/ballerina-lang/issues/24246

**Comment by KRVPerera** [comment]
Created an issue to incorporated all feedbacks related to unit test cases.
https://github.com/ballerina-platform/ballerina-lang/issues/24245

**Pull request merged**